### PR TITLE
[travis] use github token to fetch deps from ZIP files

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Is there any downside to doing so?

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
This save ~4 minutes of cumulated test time (from ~20minutes to ~16 minutes)

File: .travis.yml [hunk 1]
```diff
@@ -43,6 +43,7 @@ before_install:
     - if [[ $TRAVIS_PHP_VERSION != hhvm ]]; then echo extension = ldap.so >> $INI_FILE; fi;
     - if [[ $TRAVIS_PHP_VERSION != hhvm ]]; then phpenv config-rm xdebug.ini; fi;
     - if [[ $deps != skip ]]; then composer self-update; fi;
+    - if [[ $deps != skip ]]; then composer config -g github-oauth.github.com 52270bad1071a099c8d24629f2db2b7f07db960d; fi;
```
Comment (EmanueleMinotto) [inline .travis.yml]: Shouldn't be included in Travis/Appveyor ENVs only?

Comment (nicolas-grekas) [inline .travis.yml]: What do you mean?

**Comment by nicolas-grekas** [comment]
It also saves ~4minutes on appveyor, going from ~8 minutes to ~4 minutes

File: .travis.yml (hunk 1 above)
Comment (EmanueleMinotto) [inline .travis.yml]: That shouldn't be public or everybody will be able to use it (also in forks).

**Comment by cs278** [comment]
Might be worth commenting that the token is intended to be in there before people start reporting it.

File: .travis.yml (hunk 1 above)
Comment (nicolas-grekas) [inline .travis.yml]: good point for fork, I added a condition to prevent that

**Comment by nicolas-grekas** [comment]
@cs278 we usually don't comment the travis/appveyor files... if the concern is raised we might add it later

File: .travis.yml [hunk 2]
```diff
@@ -43,6 +44,7 @@ before_install:
     - if [[ $TRAVIS_PHP_VERSION != hhvm ]]; then echo extension = ldap.so >> $INI_FILE; fi;
     - if [[ $TRAVIS_PHP_VERSION != hhvm ]]; then phpenv config-rm xdebug.ini; fi;
     - if [[ $deps != skip ]]; then composer self-update; fi;
+    - if [[ $deps != skip && $TRAVIS_REPO_SLUG = symfony/symfony ]]; then composer config -g github-oauth.github.com 52270bad1071a099c8d24629f2db2b7f07db960d; else export COMPOSER_UPDATE='composer update --prefer-source'; fi;
```
Comment (dosten) [inline .travis.yml]: It's not possible to use a Travis env var to do not hardcode the github token?

**Comment by nicolas-grekas** [comment]
I moved the token to a dedicated file, with a comment inside

**Comment by EmanueleMinotto** [comment]
@nicolas-grekas a file just moves the problem (because people could still use the token for their own projects, with the side effect of reaching the Github API limit also for symfony/symfony).

What about the solution below?

*.travis.yml*
``` yaml
before_install:
  - composer config preferred-install source
  - 'if [[ -n "$GH_TOKEN" ]]; then composer config github-oauth.github.com ${GH_TOKEN}; fi;'
  - 'if [[ -n "$GH_TOKEN" ]]; then composer config preferred-install dist; fi;'

install:
  - composer install --no-interaction --no-progress
```

Adding a `GH_TOKEN` in Settings on https://travis-ci.org/symfony/symfony/settings (as secure, of course).

*appveyor.yml*
``` yaml
install:
  - php composer.phar config preferred-install source
  - IF %GH_TOKEN% php composer.phar config github-oauth.github.com %GH_TOKEN%
  - IF %GH_TOKEN% php composer.phar config preferred-install dist
  - php composer.phar install --no-interaction --no-progress
```

Adding a `GH_TOKEN` in Environment on https://ci.appveyor.com/project/fabpot/symfony/settings/environment (as secure too).

**Comment by nicolas-grekas** [comment]
Secure env vars are not available on PR builds, which defeat the goal of this PR.
If have absolutely no issue in making this token public. If people start ignoring the comment copying it all over, we can change it.

**Comment by dunglas** [comment]
:+1:

**Comment by stof** [comment]
:+1:

**Comment by pamil** [comment]
Hi! I'm trying to achieve the same effect on Sylius repository as pull request changing the locked dependencies are often failing because Travis cannot be authenticated against Github. I hardcoded the API key but Github noticed it and it does not work no longer. Have you encountered a similar issue?

**Comment by nicolas-grekas** [comment]
@pamil what do you mean by "I hardcoded the API key" and by "Github noticed it"? Any link? Any quote? We never experienced this.

**Comment by pamil** [comment]
@nicolas-grekas I included Github Oauth Token with `public_repo` access in Sylius' `.travis.yml`: https://github.com/pamil/Sylius/blob/28c244835443496f0b76221f26547255b3967948/.travis.yml#L42-L43

I received the following email from Github:
screen shot 2016-02-15 at 11 54 47

Seems like it's a new feature introduced after you've added the token to Symfony :)

**Comment by nicolas-grekas** [comment]
@pamil the issue is that you enabled `public_repo` access: this one gives **write** access to your public repos. You should immediately remoke this token. And create a new one **without checking any special access**. Such a token has read only access to public repos, exactly what you need I bet.

**Comment by pamil** [comment]
@nicolas-grekas oops, that's right, thank you! Github has done a great job automatically revoking that token.

**Pull request merged**